src/components/subscription-choices/subscription-choices.test.ts

```diff
@@ -0,0 +1,255 @@
+import path from 'path';
+import type { PleasantestUtils } from 'pleasantest';
+import { withBrowser, getAccessibilityTree } from 'pleasantest';
+import { loadTwigTemplate, loadGlobalCSS } from '../../../test-utils';
+
+// Helper to load the Twig template file
+const componentMarkup = loadTwigTemplate(
+  path.join(__dirname, './subscription-choices.twig')
+);
+// Helper to load the demo Twig template file
+const demoMarkup = loadTwigTemplate(path.join(__dirname, './demo/demo.twig'));
+
+// Helper to initialize the component JS
+const initJS = (utils: PleasantestUtils) =>
+  utils.runJS(`
+    import { initSubscriptionChoices } from './subscription-choices'
+    export default () => initSubscriptionChoices(
+      document.querySelector('.js-subscription-choices')
+    )
+  `);
+
+describe('Subscription Choices', () => {
+  test(
+    'should use semantic markup',
+    withBrowser(async ({ utils, page }) => {
+      await loadGlobalCSS(utils);
+      await utils.loadCSS('./subscription-choices.scss');
+      await utils.injectHTML(await componentMarkup());
+      await initJS(utils);
+
+      expect(await getAccessibilityTree(page)).toMatchInlineSnapshot(`
+        document
+          heading "Never miss an article!"
+            text "Never miss an article!"
+          status
+            text "Notifications have been turned off."
+          button "Get notifications"
+          link "Get Weekly Digests"
+            text "Get Weekly Digests"
+          form "Get Weekly Digests"
+            heading "Get Weekly Digests"
+              text "Get Weekly Digests"
+            text "Email"
+            textbox "Email"
+            button "Subscribe"
+      `);
+    })
+  );
+
+  test(
+    'should be keyboard accessible',
+    withBrowser(async ({ utils, screen, waitFor, page }) => {
+      await loadGlobalCSS(utils);
+      await utils.loadCSS('./subscription-choices.scss');
+      await utils.injectHTML(await demoMarkup());
+      await initJS(utils);
+
+      // Helper function used throughout this test
+      //
+      // I eslint-disabled because it was wanted me to pull this function
+      // outside of the scope of this test, but, I feel collocating the function
+      // within the only test that uses the function makes more sense.
+      //
+      // eslint-disable-next-line @cloudfour/unicorn/consistent-function-scoping
+      const getFormDimensions = (formEl: HTMLElement) => ({
+        formHeight: formEl.clientHeight,
+        formWidth: formEl.clientWidth,
+      });
+
+      // Confirm the form is visually hidden by default
+      const form = await screen.getByRole('form', {
+        name: 'Get Weekly Digests',
+      });
+      let { formHeight, formWidth } = await form.evaluate(getFormDimensions);
+      expect(formHeight).toBeLessThanOrEqual(1);
+      expect(formWidth).toBeLessThanOrEqual(1);
```

Comment:
@gerardo-rodriguez Is there a reason we can't use [`.toBeVisible()`](https://github.com/testing-library/jest-dom#tobevisible) here?

Reply:
Great question, @calebeby!

The form is _technically_ visible so that assertion fails. The reason it is still visible is that it uses the [`@include a11y.sr-only` mixin](https://github.com/cloudfour/cloudfour.com-patterns/pull/1679/files#diff-feb19a35856697c41225bad7dda4443f13f7ccab3052b5c9a15bb1d2203cc10fR58) to ensure the form is accessible via keyboard navigation. It is only _visually_ hidden.

Reply:
The toBeVisible matcher is supposed to represent visual visibility. So .not.toBeVisible should pass even if for example an element has `opacity: 0` (so it'd still be in the accessibility tree), otherwise there is a bug with .not.toBeVisible. Mind opening an issue?

Reply:
@calebeby These are the values for the [`a11y.sr-only` mixin](https://github.com/cloudfour/cloudfour.com-patterns/blob/v-next/src/mixins/_a11y.scss):
```css
  block-size: 1px,
  border: 0,
  clip: rect(0 0 0 0),
  clip-path: polygon(0 0, 0 0, 0 0),
  inline-size: 1px,
  margin: -1px,
  overflow: hidden,
  padding: 0,
  position: absolute,
  white-space: nowrap,
```
I don't think it's a bug, the `opacity` is not set to `0`. The form is _visible_ to assistive technologies like a screen reader. If you tab into the form, the form will be visually visible as well. This is the same behavior as the "Skip to main content" link.

<img width="481" alt="Screen Shot 2022-03-22 at 3 12 11 PM" src="https://user-images.githubusercontent.com/459757/159584974-e897bfd1-a233-4140-bef3-b251c46fc296.png">

Reply:
I still think it is a bug. opacity:0 is also still visible to assistive technologies too (at least based on what Chrome devtools shows). My understanding of the toBeVisible matcher is that it is supposed to check that something is visually visible, not that it is present in the accessibility tree.

Reply:
> My understanding of the toBeVisible matcher is that it is supposed to check that something is visually visible, not that it is present in the accessibility tree.

@calebeby Interesting! I hadn't thought about it.

I did a quick search and found [this](https://github.com/testing-library/jest-dom/issues/338#issuecomment-878882097) (off of your GH issue 🙃):

> In testing-library we usually go with "visible" = possibly perceivable + accessible by assistive technology. E.g. display: none is impossible to be perceivable and inaccessible. On the other hand zero-dimensions is impossible to be perceivable but still be considered visible because the element would be exposed to assistive technology.

If we subscribe to that, then `toBeVisible()` is doing what it's supposed to be doing. 🙂

Reply:
It seems like based on that discussion, the zero-dimensions case is chosen because of the limitations of JSDOM, where elements don't have size.

In pleasantest, we don't have that constraint, we _could_ make the behavior match the zero-opacity case where the element is accessible by assistive technology but not possibly perceivable. In the zero-opacity case, the element is not considered visible by jest-dom. We could make pleasantest also consider the zero-size case to not be visible.

:shrug: Probably a discussion that can happen outside this PR.

In this PR, would it make sense to separate out a `expectElementToHaveSize` function?

```js
await expectElementToHaveSize(form);
```

```js
const expectElementToHaveSize = async () => {
	const { formHeight, formWidth } = await form.evaluate((formEl: HTMLElement) => ({
		formHeight: formEl.clientHeight,
		formWidth: formEl.clientWidth,
	}));
	expect(formHeight).toBeGreaterThan(1);
	expect(formWidth).toBeGreaterThan(1);
}
```

Reply:
To me this line:
```js
await expectElementToHaveSize(form);
```
is more clear about what it is doing than these lines:
```js
({ formHeight, formWidth } = await form.evaluate(getFormDimensions));
expect(formHeight).toBeGreaterThan(1);
expect(formWidth).toBeGreaterThan(1);
```

I don't feel super strongly about this.

Reply:
@calebeby That's a fair argument.

Do I create a separate function for the `toBeLessThanOrEqual(1)` version?

One is checking:
```js
 expect(formHeight).toBeGreaterThan(1);
expect(formWidth).toBeGreaterThan(1);
```
The other is checking:
```js
expect(formHeight).toBeLessThanOrEqual(1);
expect(formWidth).toBeLessThanOrEqual(1);
```
The element is expected to have a size in both cases. (`1` or greater than one). Which this in mind, I'm not sure `expectElementToHaveSize()` would work as nicely.

Hmm... 🤔

Reply:
@calebeby I've made changes! https://github.com/cloudfour/cloudfour.com-patterns/pull/1679/commits/c9216fdc855dc8c940c79c983e280325d7b069ff